### DISCUSSION TITLE
I2P Simulation ending early

### DISCUSSION THREAD

**Discussion by sxlissxn** [shadow/shadow#3398] [Q&A]
Hello, I am trying to conduct an I2Pd network simulation, where 2 routers communicate with each other. The first one need to run at least one min before the 2nd router so that the connection can be made. However, they don't seem to be coordinated. I'm  not sure if its incorrect configuration or Is the shadow network simulator the ideal tool for this simulation?

**Comment by sporksmith** [comment]
I think [hosts.<hostname>.processes[*].start_time](https://shadow.github.io/docs/guide/shadow_config_spec.html#hostshostnameprocessesstart_time) is what you want.

If you still need help, please share the shadow config file that you're using.

**Comment by sxlissxn** [comment]
I've used ` start_time: 1 min` for it but i still get this anomaly.
Here's the yaml file:
```yaml
general:
  model_unblocked_syscall_latency: true
  stop_time: 30 min
network:
  graph:
    type: gml
    inline: |
      graph [
        directed 0
        node [
          id 0
          host_bandwidth_down "1 Gbit"
          host_bandwidth_up "1 Gbit"
        ]
        node [
          id 1
          host_bandwidth_down "1 Gbit"
          host_bandwidth_up "1 Gbit"
        ]
        edge [
          source 0
          target 0
          latency "50 ms"
          jitter "0 ms"
          packet_loss 0.0
        ]
        edge [
          source 1
          target 0
          latency "50 ms"
          jitter "0 ms"
          packet_loss 0.0
        ]
        edge [
          source 0
          target 1
          latency "50 ms"
          jitter "0 ms"
          packet_loss 0.0
        ]
        edge [
          source 1
          target 1
          latency "50 ms"
          jitter "0 ms"
          packet_loss 0.0
        ]
      ]
hosts:
  
  r1:
    network_node_id: 0
    ip_addr: 100.0.0.1
    processes:
    - path: i2pd
      args: 
        --conf /.local/shadow/i2psimulation/i2psim/conf/r1/i2pd.conf
        --datadir /.local/shadow/i2psimulation/i2psim/conf/r1
      
      start_time: 1 min
      expected_final_state: running
 
  r2:
    network_node_id: 0
    ip_addr: 100.0.0.2
    processes:
    
    - path: i2pd
      args: 
         --conf /.local/shadow/i2psimulation/i2psim/conf/r2/i2pd.conf
         --datadir /.local/shadow/i2psimulation/i2psim/conf/r2
      start_time: 3 min
      expected_final_state: running
    - path: tgen
      environment: { OPENBLAS_NUM_THREADS: "1" }
      args: /.local/shadow/i2psimulation/i2psim/conf/tgen.client.graphml.xml
```

**Reply by robgjansen** [reply]
It appears that you have not specified a `start_time` for tgen on host r2, meaning that tgen would start at `start_time=0` by default, which is _before_ the time that i2pd is started. Thus, if the tgen.client.graphml.xml is configured to connect through the r2 i2pd socks proxy, tgen would likely fail because the proxy would not be running yet. Is that the desired behavior?

Otherwise, we would need more information about what seems to be failing to help. You can grep through the shadow.log file or some of the process log files to check for warnings and errors. See for example:
- Shadow log file: https://shadow.github.io/docs/guide/log_format.html
- Process log files: https://shadow.github.io/docs/guide/getting_started_basic.html#viewing-the-simulation-output

**Comment by sxlissxn** [comment]
Hello, the simulation is subject to errors as the i2pd routers doesn't connect despite both routers having netDB information about each other:
![image](https://github.com/user-attachments/assets/4fd7e30f-277e-4a14-a3ca-945318baab8f)
The shadow logs seems fine, but the tgen streams are failing:
![image](https://github.com/user-attachments/assets/ef670239-5b91-40c0-ae15-8ad1e3286ca4)

**Reply by sporksmith** [reply]
It seems like an i2p configuration issue. Processes inside shadow can't communicate with the outside world; you need to set up a private i2p network inside the shadow simulation s.t. the i2p routers know about eachother and *don't* try to use the real public i2p network. If you're already doing that, you might just need more i2p relays inside the simulation - with only 2 relays that are trying to communicate with eachother I don't think they'll be able to set up a multihop tunnel.

I would look for i2p documentation/support for bootstrapping a private i2p network. I didn't find a whole lot myself from a quick search, though this might be relevant: https://www.reddit.com/r/i2p/comments/76uael/how_does_i2p_bootstrap/

**Comment by sxlissxn** [comment]
hello, the simulation is encountering this issue:
![image](https://github.com/user-attachments/assets/fca598f1-9150-49c8-ba85-47a0bebe89fd)
I have not changed the IP configurations, only added the keys and certificates.

**Reply by stevenengler** [reply]
The issue is that one of your programs is sending a packet to an IP address that's not part of the simulation. Shadow should ideally just drop the packet (see issue #3148) instead of panicking. But in any case, sending a packet to an IP that's not part of the simulation probably isn't what you want, and possibly means that your simulation isn't configured correctly. In this case, it's probably because you haven't changed the IP configuration.

**Reply by stevenengler** [reply]
#3148 is fixed now, so you shouldn't see this panic anymore and will just see a warning in the log about the dropped packet.
